crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1560,6 +1753,44 @@ impl<'db> Type<'db> {
         arguments: &CallArguments<'_, 'db>,
     ) -> Result<CallOutcome<'db>, CallError<'db>> {
         match self {
+            Type::Callable(CallableType::BoundMethod(bound_method)) => {
+                let instance = bound_method.self_instance(db);
+                let arguments = arguments.with_self(instance);
+
+                let binding = bind_call(
+                    db,
+                    &arguments,
+                    bound_method.function(db).signature(db),
+                    self,
+                );
+
+                if binding.has_binding_errors() {
+                    Err(CallError::BindingError { binding })
+                } else {
+                    Ok(CallOutcome::Single(binding))
+                }
+            }
+            Type::Callable(CallableType::MethodWrapperDunderGet(function)) => {
+                let return_ty = match arguments.first_argument() {
+                    Some(ty) if ty.is_none(db) => Type::FunctionLiteral(function),
+                    Some(instance) => Type::Callable(CallableType::BoundMethod(
+                        BoundMethodType::new(db, function, instance),
+                    )),
```

Comment:
Should this error if the first argument is `None`?

Reply:
No, first argument being `None` is the signal that the descriptor is being called on a class instead of an instance, it's not an error; returning the function itself is correct in that case (for a function `__get__`) and matches the runtime behavior.

Reply:
Yes, thank you! I'm now modelling this much more precisely by creating a `Signature` dynamically, which let's us catch all kinds of call errors. This also revealed two minor mistakes in one of the descriptor tests where I was passing a function `f` as the `owner` argument, instead of `type(f)`. Check out the changes in this commit: 28e37cbbea957b03cafbebda761dba2815a61699

Reply:
> No, first argument being `None` is the signal that the descriptor is being called on a class instead of an instance, it's not an error

I think we need to distinguish Rust's `None` and Python's `None`. We handle a Python `None` correctly already. What @MichaReiser was referring to was probably a Rust `None`, i.e. that no parameter is passed at all. This was *not* handled correctly, but it is now fixed with the reference commit.